scio-extra/src/main/scala/com/spotify/scio/extra/voyager/VoyagerUri.scala

```diff
@@ -0,0 +1,166 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra.voyager
+
+import com.spotify.scio.coders.Coder
+import com.spotify.scio.util.{RemoteFileUtil, ScioUtil}
+import com.spotify.voyager.jni.Index
+import com.spotify.voyager.jni.Index.{SpaceType, StorageDataType}
+import org.apache.beam.sdk.options.PipelineOptions
+
+import java.io.File
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Path, Paths}
+import scala.collection.mutable
+
+/**
+ * Represents the base URI for a voyager index, either on a local or a remote file system. For
+ * remote file systems, the `path` should be in the form 'scheme://<bucket>/<path>/'. For local
+ * files, it should be in the form '/<path>/'. The `path` specified represents the directory where
+ * the `index.hnsw` and `names.json` are.
+ */
+sealed trait VoyagerUri {
+  def path: String
+  private[voyager] def getReader(
+    distanceMeasure: SpaceType,
+    storageDataType: StorageDataType,
+    dim: Int
+  ): VoyagerReader
+  private[voyager] def saveAndClose(voyagerWriter: VoyagerWriter): Unit
+  private[voyager] def exists: Boolean
+}
+
+private[voyager] object VoyagerUri {
+  def apply(path: String, opts: PipelineOptions): VoyagerUri = {
+    if (ScioUtil.isLocalUri(new URI(path))) {
+      LocalVoyagerUri(path)
+    } else {
+      val rfu: RemoteFileUtil = RemoteFileUtil.create(opts)
+      RemoteVoyagerUri(path, rfu)
+    }
+  }
+
+  def files: Seq[String] = Seq("index.hnsw", "names.json")
+}
+
+case class LocalVoyagerUri(path: String) extends VoyagerUri {
+  override private[voyager] def getReader(
+    distanceMeasure: SpaceType,
+    storageType: StorageDataType,
+    dim: Int
+  ): VoyagerReader = {
+
+    val indexFileName: String = path + "/index.hnsw"
+    val namesFileName: String = path + "/names.json"
+    new VoyagerReader(indexFileName, namesFileName, distanceMeasure, storageType, dim)
+  }
+
+  override private[voyager] def saveAndClose(w: VoyagerWriter): Unit = {
+    w.save(path)
+    w.close()
+  }
+
+  override private[voyager] def exists: Boolean =
+    VoyagerUri.files.exists(f => new File(path + "/" + f).exists())
+}
+
+case class RemoteVoyagerUri(
+  path: String,
+  remoteFileUtil: RemoteFileUtil
+) extends VoyagerUri {
+  override private[voyager] def getReader(
+    distanceMeasure: SpaceType,
+    storageType: StorageDataType,
+    dim: Int
+  ): VoyagerReader = {
+    val indexFileName: String = remoteFileUtil.download(new URI(path + "/index.hnsw")).toString
+    val namesFileName: String = remoteFileUtil.download(new URI(path + "/names.json")).toString
+    new VoyagerReader(indexFileName, namesFileName, distanceMeasure, storageType, dim)
+  }
+
+  override private[voyager] def saveAndClose(w: VoyagerWriter): Unit = {
+    val tempPath: Path = Files.createTempDirectory("")
+    logger.info(s"temp path: $tempPath")
```

Comment:
```suggestion
```

I would move this to line 106, so we can log it as "Uploaded Voyager file from $tempPath to  $path/$f" 👍